### PR TITLE
convert optional columms to float by default [WIP]

### PR DESCRIPTION
**Pull request by PGryllos**
not test data or test included ftm

### CODE REVIEW

**Comment by PGryllos** [comment]
cc: @martindurant

File: fastparquet/api.py [hunk 1]
```diff
@@ -497,17 +497,20 @@ def _dtypes(self, categories=None):
...
-                        break
-                    if chunk.meta_data.statistics.null_count is None:
-                        num_nulls = True
-                        break
-                    if chunk.meta_data.statistics.null_count:
-                        num_nulls = True
-                        break
+                if not self.schema.is_required(col):
+                    num_nulls = True
+                else:
```
Comment (martindurant) [inline fastparquet/api.py]: In fact, the num_nulls tests below will never run for now for any column that could have nulls, so in this version it is wasted code. 

The initial check could be `'.' in col and not self.schema.is_required(col)`, i.e., we only act so pessimistically in this special nested case, but keep the old behaviour for normal columns. This means that an optional int column will stay as int in the case that we can be certain it has no nulls.

File: setup.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def fix_exts(sources):
 
 setup(
     name='fastparquet',
-    version='0.1.6',
+    version='0.1.6.feature.optional_columns_to_float.0',
```
Comment (martindurant) [inline setup.py]: There is no need to do this, I will take care of the version number for release after the PR is merged.
Most of my projects have moved to versioneer, which automatically adds git information to the version number, but not this one.

Comment (PGryllos) [reply setup.py]: sorry, did that for my internal build pipeline; need to remove it?

Comment (martindurant) [reply setup.py]: Yes, I think its best if the version number only changes at times of release.

**Comment by martindurant** [comment]
Fixes #373

File: setup.py (hunk 2 above)
Comment (PGryllos) [reply setup.py]: want me to re-write git history or just remove it?

Comment (martindurant) [reply setup.py]: Just remove is fine.

Comment (PGryllos) [reply setup.py]: done, need to check later why the tests fail

**Comment by PGryllos** [comment]
@martindurant I am looking into the failing unit test

```python
    def test_dotted_column(tempdir):
        fn = os.path.join(tempdir, 'tmp.parq')
        df = pd.DataFrame({'x.y': [1, 2, 3]
                           'y': [1., 2., 3.]})
    
        writer.write(fn, df)
    
>       out = ParquetFile(fn).to_pandas()
```

there during reading the introduced is_required check breaks when trying to find the children node x. I am not sure what is exactly the problem. According to the semantics that I have understood so far the is_required field should not break for anything unless the field that we are looking for does not exist. When trying with the column x.y the `root.children[x]' fails. The existing children at this point are x.y and y. I cannot judge if this correct or not. If the field x not part of the children justifiably because it's the only parent?

**Comment by PGryllos** [comment]
> However, we could use something more rare, like '$$' or '_sub_'

so are you suggesting to replace the dot with one of those throughout the codebase to denote hierarchy?

**Comment by martindurant** [comment]
(note that I had put underscore characters around "sub" in my suggestion, but it didn't render)

You would need to look carefully in the code for for `'.'.join()` and `split('.')` as well as for cases like I was suggesting here, `if '.' in..`. The majority will be in schema and api modules, exactly the parts of code we are talking about.

**Comment by martindurant** [comment]
The point is, that columns names *can* contain '.' characters, which is exactly what happens in the test which is now failing. 
So either we need
- a better, more unique separator to test on,
- or to rewrite the logic in things like `is_required()` to explicitly walk the schema tree upwards (starting with, in `schema_tree()` after line `root.children[s.name] = s`, adding `s.parent = root`).

I suggest the former is far easier. I find 10 instances of `'.'` and 8 of `"."` in the codebase, some of which are file paths or in tests.

The final alternative, perhaps simpler still, is to figure out what you changed that makes `test_dotted_column` fail. 

-EDIT-
I have the answer: you are currently using the string `col` in `is_required`, and splitting this on the delimiter is the problem. However, each chunk actually has the full schema path as a list, so if you call `is_required(chunk.metadata.path_in_schema)` you will not have this problem; and indeed, then the test should be:
```
path = chunk.metadata.path_in_schema
if len(path) > 1 and not self.schema.is_required(path[:-1]):
    num_nulls = True
    break
```
(which, obviously, would need to be within the row-group loop, but would short-circuit).

**Comment by PGryllos** [comment]
@martindurant thx a lot for looking into this; I will try out your suggestion. fyi I figured a way to avoid loading the complicated nested structure for the moment by adding one more etl step on the spark pipeline and then go from a simple structure of serialised objects (also asked on SO about it https://stackoverflow.com/questions/53169690/is-there-any-good-way-to-read-the-content-of-a-spark-rdd-into-a-dask-structure) 

I am still very much interested in being able to load the nested structure directly because it makes integration much more seamless; but a quick inspection of the loaded structure (with the patched code that currently breaks the tests) showed that even if there were no errors during loading some of the nested columns were actually missing; So I need to figure out why that is. I remember you mentioning that nested structures except a few are not supported; Where can I look to get a good grip of what is supported? Also is the long term plan to eventually support nested structures completely or not? We are generally moving towards simplifying our structures :'D but good to know what your plans are for the future

again thx for the help

**Comment by martindurant** [comment]
https://fastparquet.readthedocs.io/en/latest/details.html#reading-nested-schema is still the best reference. And no, there is no plan to support anything complex than that. As I think I said early on, there should be support for this in arrow and hopefully the pyarrow wrapping; but you could also try the alpha-stage awkward-array package if you want to experiment.

Note also, that the parquet spec includes JSON/BSON encoding of objects in columns, and fastparquet supports this. Unfortunately, spark does not...

**Comment by eddy-geek** [comment]
I am also hitting `ValueError: cannot convert float NaN to integer`
But since this pR, Pandas 0.24 added a [Nullable Integer Type](https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html), isn't using it better than using f8 floats?

**Comment by martindurant** [comment]
I agree, it would be better, if only someone would take the time to implement it!

**Pull request closed without merging**